File: Leetcode/maxRunningTimeOfNComputers.cpp
#include <bits/stdc++.h>

using namespace std;

bool isPossible(vector<int> &batteries, long long time, int computers)
{
  long long totTime = 0;
  for (long long bTime : batteries)
    totTime += min(bTime, time);
  return (totTime >= time * computers);
}

long long maxRunTime(int n, vector<int> &batteries)
{
  long long low = 1, high = 1e14;
  while (low <= high)
  {
    long long mid = low + (high - low) / 2;
    if (isPossible(batteries, mid, n))
      low = mid + 1;
    else
      high = mid - 1;
  }
  return high;
}

int main()
{
  int n, battery;
  cin >> n >> battery;
  vector<int> batteries(battery);
  for (int i = 0; i < battery; i++)
    cin >> batteries[i];

  cout << maxRunTime(n, batteries) << "\n";

  return 0;
}

File: Leetcode/Python/max_running_time_of_n_computers.py
def main():
  n = int(input())
  batteries = list(map(int, input().split()))
  print(maxRunTime(n, batteries))

def canRun(batteries: list[int], n: int, time: int) -> bool:
  total = 0
  for battery in batteries:
    total += min(battery, time)
    if total >= n * time:
      return True
  return total >= n * time

def maxRunTime(n: int, batteries: list[int]) -> int:
  low , high = 1, 10**14
  while low <= high:
    mid = low + (high - low) // 2
    if canRun(batteries, n, mid):
      low = mid + 1
    else:
      high = mid - 1
  return high

if __name__ == '__main__':
  main()